wasm/src/lsp/types.rs

```diff
@@ -0,0 +1,300 @@
+    if host.is_some() {
+        return Err(ToFilePathErr);
+    }
+
+    let mut bytes = Vec::new();
+
+    for segment in segments {
+        bytes.push(b'/');
+        bytes.extend(percent_decode(segment.as_bytes()));
+    }
+
+    // A windows drive letter must end with a slash.
+    if bytes.len() > 2
+        && matches!(bytes[bytes.len() - 2], b'a'..=b'z' | b'A'..=b'Z')
+        && matches!(bytes[bytes.len() - 1], b':' | b'|')
+    {
+        bytes.push(b'/');
+    }
+
+    let path = PathBuf::from(decode_string(&bytes));
+
+    debug_assert!(
+        path.is_absolute(),
+        "to_file_path() failed to produce an absolute Path"
+    );
+
+    Ok(path)
+}
+
+impl HasFilePath for Url {
+    fn our_to_file_path(&self) -> Result<PathBuf, ToFilePathErr> {
+        if let Some(segments) = self.path_segments() {
+            let host = match self.host() {
+                None | Some(Host::Domain("localhost")) => None,
+                Some(h) if self.scheme() == "file" => Some(h.to_string().as_bytes().to_owned()),
+                _ => return Err(ToFilePathErr),
+            };
+
+            return file_url_segments_to_pathbuf(host, segments);
+        }
+        Err(ToFilePathErr)
+    }
+}
+
+#[cfg(test)]
+fn uniterr<A>(_: A) -> () { () }
+
+#[test]
+fn test_file_segments_to_pathbuf_1() {
+    assert_eq!(
+        Url::parse("fink:::::not/good").map_err(uniterr).and_then(|uri| {
+            uri.our_to_file_path().map_err(uniterr)
+        }),
+        Err(())
+    );
+}
+
+#[test]
+fn test_file_segments_to_pathbuf_2() {
+    assert_eq!(
+        Url::parse("file:///home/person/stuff.txt").map_err(uniterr).and_then(|uri| {
+            uri.our_to_file_path().map_err(uniterr)
+        }),
+        PathBuf::from_str("/home/person/stuff.txt").map_err(uniterr)
+    );
+}
+
+pub fn cast<R>(req: Request) -> Result<(RequestId, R::Params), ExtractError<Request>>
+where
+    R: lsp_types::request::Request,
+    R::Params: serde::de::DeserializeOwned,
+{
+    req.extract(R::METHOD)
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct DocPosition {
+    pub line: u32,
+    pub character: u32,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct DocRange {
+    pub start: DocPosition,
+    pub end: DocPosition,
+}
+
+#[derive(Clone, Debug)]
+pub struct DocPatch {
+    pub range: DocRange,
+    pub text: String,
+}
+
+impl DocPosition {
+    // This is not dead code, but due to the variation in build configuration
+    // not all ways of building this module require it.
+    // among test, rustlib and wasm32.
+    #[allow(dead_code)]
+    pub fn from_position(pos: &Position) -> Self {
+        DocPosition {
+            line: pos.line,
+            character: pos.character,
+        }
+
+    }
+
+    pub fn to_position(&self) -> Position {
+        Position {
+            line: self.line,
+            character: self.character,
+        }
+    }
+}
+
+impl DocRange {
```

Comment:
This could use a comment about `DocRange` and Srcloc being zero or one-based.

Reply:
Ya will add.